mpqp/tools/maths.py

```diff
@@ -307,47 +351,71 @@ def rand_unitary_2x2_matrix() -> npt.NDArray[np.complex64]:
     return np.array([[c, -eg * s], [eg * s, eg * ep * c]])
 
 
-def rand_product_local_unitaries(nb_qubits: int) -> npt.NDArray[np.complex64]:
+def rand_product_local_unitaries(
+    nb_qubits: int, seed: Optional[Union[int, np.random.Generator]] = None
+) -> npt.NDArray[np.complex64]:
     """Generate a pseudo random matrix, resulting from a tensor product of
     random unitary matrices.
 
     Args:
         nb_qubits: Number of qubits on which the product of unitaries will act.
+        seed: Seed used to control the random generation of the matrix.
 
     Returns:
         A tensor product of random unitary matrices.
 
     Example:
         >>> rand_product_local_unitaries(2) # doctest: +SKIP
-        array([[-0.39648015+0.j        ,  0.49842218-0.16609181j,
-                   0.39826454-0.21692223j, -0.40979321+0.43953607j],
-               [-0.49842218+0.16609181j,  0.14052896-0.37073997j,
-                   0.40979321-0.43953607j,  0.06167784+0.44929471j],
-               [-0.39826454+0.21692223j,  0.40979321-0.43953607j,
-                   0.16112375-0.36226461j, -0.05079312+0.52290651j],
-               [-0.40979321+0.43953607j, -0.06167784-0.44929471j,
-                   0.05079312-0.52290651j,  0.28163685+0.27906487j]])
+        array([[ 0.91443498+0.j        , -0.34882095-0.2052623j ],
+               [ 0.34882095+0.2052623j , -0.00785861+0.91440121j]])
+
+        >>> rand_product_local_unitaries(2, seed=123)
+        array([[-0.54205051+0.j        , -0.15559823-0.82581501j],
+               [ 0.15559823+0.82581501j,  0.08203889-0.53580629j]])
 
     """
-    return reduce(np.kron, [rand_unitary_2x2_matrix() for _ in range(nb_qubits - 1)])
+    if seed is None:
+        rng = np.random.default_rng()
+    elif isinstance(seed, np.random.Generator):
+        rng = seed
+    else:
+        rng = np.random.default_rng(seed)
+
+    return reduce(
+        np.kron, [rand_unitary_2x2_matrix(seed=rng) for _ in range(nb_qubits - 1)]
```

Comment:
here we have a problem. If the user gives a seed for the product of local unitaries, he doesn't necessarly expected that all the 2x2 unitarties will have the same seed and will be identical.

I suggest that we find a way to decompose uniquely the seed given in parameter to multiple seed that will be passed one by one for each rand_unitary 2x2 matrix generation. (maybe use prime decomposition of the integer, and put 1 as a seed if we don't have enough factors)

Reply:
I think here we could implment seed decomposition using incrementing values, so the overall seed generation is controlled by the initial user provided seed